aws/apis/identity/v1alpha1/iamrole_helpers.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+
+	runtimev1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+)
+
+// SetBindingPhase of this IAMRole.
+func (r *IAMRole) SetBindingPhase(p runtimev1alpha1.BindingPhase) {
+	r.Status.SetBindingPhase(p)
+}
+
+// GetBindingPhase of this IAMRole.
+func (r *IAMRole) GetBindingPhase() runtimev1alpha1.BindingPhase {
+	return r.Status.GetBindingPhase()
+}
+
+// SetConditions of this IAMRole.
+func (r *IAMRole) SetConditions(c ...runtimev1alpha1.Condition) {
+	r.Status.SetConditions(c...)
+}
+
+// SetClaimReference of this IAMRole.
+func (r *IAMRole) SetClaimReference(ref *corev1.ObjectReference) {
+	r.Spec.ClaimReference = ref
+}
+
+// GetClaimReference of this IAMRole.
+func (r *IAMRole) GetClaimReference() *corev1.ObjectReference {
+	return r.Spec.ClaimReference
+}
+
+// SetClassReference of this IAMRole.
+func (r *IAMRole) SetClassReference(ref *corev1.ObjectReference) {
+	r.Spec.ClassReference = ref
+}
+
+// GetClassReference of this IAMRole.
+func (r *IAMRole) GetClassReference() *corev1.ObjectReference {
+	return r.Spec.ClassReference
+}
+
+// SetWriteConnectionSecretToReference of this IAMRole.
+func (r *IAMRole) SetWriteConnectionSecretToReference(ref corev1.LocalObjectReference) {
+	r.Spec.WriteConnectionSecretToReference = ref
+}
+
+// GetWriteConnectionSecretToReference of this IAMRole.
+func (r *IAMRole) GetWriteConnectionSecretToReference() corev1.LocalObjectReference {
+	return r.Spec.WriteConnectionSecretToReference
+}
+
+// GetReclaimPolicy of this IAMRole.
+func (r *IAMRole) GetReclaimPolicy() runtimev1alpha1.ReclaimPolicy {
+	return r.Spec.ReclaimPolicy
+}
+
+// SetReclaimPolicy of this IAMRole.
+func (r *IAMRole) SetReclaimPolicy(p runtimev1alpha1.ReclaimPolicy) {
+	r.Spec.ReclaimPolicy = p
+}
+
+// BuildExternalStateFromObservation builds the external resource state, using the given observation
+func (r *IAMRole) BuildExternalStateFromObservation(observation iam.Role) IAMRoleExternalState {
```

Comment:
Good catch! Initially this was updating the same object (instead of building an state object), and that's why it's there. It is no longer needed to be an object method.

Reply:
However, it seems that making this a standalone method would cause naming conflicts with the other resource type in the same library. (e.g. `BuildExternalStateFromObservation` of an `IAMRole` will conflict with that of `IAMRolePolicyAttachment`).

Do you think it's worth make these objects static, and then resolve the conflict by adding the type name to it? This would create long method names like `BuildIAMRolePolicyAttachmentExternalStateFromObservation`

Reply:
https://github.com/crossplaneio/crossplane/pull/750/files#diff-f452c510bde3dbe33c45b8baaa011ac6R68

My personal preference is to avoid having this "translation" code alongside the API type definitions, in order to keep our definitions a little less tightly coupled to (and dependent upon, linker wise) the cloud provider API objects they correspond with. The above link contains an example of the pattern I'm currently using.

Another option (that I am personally less fond of) would be to keep these as methods on the underlying types that modify them rather than replacing or returning them. This could be a method named something like `UpdateFrom(observation iam.Role)` on `IAMRoleExternalState`, or a method named `UpdateExternalStateFrom(observation iam.Role)` on `IAMRole`.

Reply:
I actually think this logic should reside in the API type definitions, as it is dealing with Group/Version specific field definitions. 
As for returning a status instance, rather than updating the object itself I wanted to make it clear that all status modification are actually happening in the controller itself, rather than doing a call which recursively modifies the object itself.

Reply:
> this was the confusion that was causing the error I was getting regarding non-stop reconciliation loop

When you were experiencing this issue I believe it was because you were _replacing_ the status, e.g.:

```go
func (r *IAMRole) UpdateStatus(observed *iam.Role) {
    r.Status = IAMRoleStatus{
        SomeField = observed.SomeField,
    }
} 
```

This was an issue because some fields that are common to all managed resource statuses (i.e. the Conditions array) contain logic to avoid unnecessary churn when updated (i.e. avoiding updating their transition timestamp if the condition is already set).

Now that `IAMRoleExternalStatus` is a distinct embedded field of `IAMRoleStatus`, I believe it should be safe to either set it or update it in place. i.e.:

```go
func (r *IAMRole) UpdateStatus(observed *iam.Role) {
    r.Status.IAMRoleExternalStatus = IAMRoleExternalStatus{
        SomeField = observed.SomeField,
    }
} 
```

or

```go
func (r *IAMRole) UpdateStatus(observed *iam.Role) {
    r.Status.SomeField = observed.SomeField
    r.Status.SomeOtherField = observed.SomeOtherField
} 
```

I'd prefer either of the above to maintaining a method that does not use its receiver. I'd also prefer a function (not method) in this package with a cumbersome name over the current pattern. I believe this is the only outstanding comment thread, so I've approved the PR and will let you decide how to proceed.